### PR TITLE
RFC: User-friendly Configurable Connectors

### PR DESCRIPTION
**Pull request by Velfi**
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: design/src/SUMMARY.md [hunk 1]
```diff
@@ -27,3 +27,4 @@
   - [RFC-0011: Publishing Alpha to Crates.io](./rfcs/rfc0011_crates_io_alpha_publishing.md)
   - [RFC-0012: Independent Crate Versioning](./rfcs/rfc0012_independent_crate_versioning.md)
   - [RFC-0013: Body Callback APIs](./rfcs/rfc0013_body_callback_apis.md)
+  - [RFC-0015: User-friendly Configurable Connectors](./rfcs/rfc0002_http_versions.mdrfc0015_user_friendly_configurable_connectors)
```
Comment (jdisanti) [inline design/src/SUMMARY.md:30]:
```suggestion
  - [RFC-0015: User-friendly Configurable Connectors](./rfcs/rfc0015_user_friendly_configurable_connectors.md)
```

Comment (Velfi) [reply design/src/SUMMARY.md:30]: derp

File: design/src/rfcs/rfc0015_user_friendly_configurable_connectors.md [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+
+1. The Fluent Builder constructs the inner Operation, passing it a copy of the Service Config and the Smithy Client Pool.
+2. All operations have, in the HTTP request Property Bag, a priority list of HTTP versions to use when making requests. That list is fetched from the property bag.
+3. A `MakeConnectorSettings` is created by calling an `Into::into` impl on the Service Config.
+4. Using those settings, and the priority list of HTTP versions, a client is fetched from the Smithy Client Pool.
+5. Finally, the request is sent by calling `client.call(operation).await`
+
+### `let bucket_list = resp.buckets().unwrap();`
+
+There's nothing new to be seen here. The previous line sent the request and recieved the response. This line is just looking at the response.
```
Comment (jdisanti) [inline design/src/rfcs/rfc0015_user_friendly_configurable_connectors.md:93]:
```suggestion
There's nothing new to be seen here. The previous line sent the request and received the response. This line is just looking at the response.
```

File: design/src/rfcs/rfc0015_user_friendly_configurable_connectors.md [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+
+- `pub async fn get_or_create_client<E>(&self, make_connector_settings: MakeConnectorSettings, http_versions: &[http::Version]) -> Result<Arc<aws_smithy_client::Client>, SdkError<E>>`
+- `async fn fetch_existing_client(&self, connector_key: &ConnectorKey) -> Option<Arc<aws_smithy_client::Client>>`
+- `async fn initialize_and_store_new_client<E>(&self, connector_key: &ConnectorKey) -> Result<Arc<aws_smithy_client::Client>, SdkError<E>>`
+
+Let's look at each in turn, starting with the sole public method.
+
+### `get_or_create_client<E>`
+
+Smithy Clients are created lazily. If a user never makes a request, than no clients get created. If they only make `HTTP/1.1` requests, then only an `HTTP/1.1` is ever created. When users call `send()` on an Operation, that in turn calls this method. Here's what the implementation could look like:
```
Comment (jdisanti) [inline design/src/rfcs/rfc0015_user_friendly_configurable_connectors.md:108]:
```suggestion
Smithy Clients are created lazily. If a user never makes a request, then no clients get created. If they only make `HTTP/1.1` requests, then only an `HTTP/1.1` is ever created. When users call `send()` on an Operation, that in turn calls this method. Here's what the implementation could look like:
```

File: design/src/rfcs/rfc0015_user_friendly_configurable_connectors.md [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+There's nothing new to be seen here. The previous line sent the request and recieved the response. This line is just looking at the response.
+
+The inner workings of the Smithy Client Pool
+--------------------------------------------
+
+While the Service Client "owns" the Smithy Client Pool, its only responsibility is to share copies of it with Operations. The client pool has three methods:
+
+- `pub async fn get_or_create_client<E>(&self, make_connector_settings: MakeConnectorSettings, http_versions: &[http::Version]) -> Result<Arc<aws_smithy_client::Client>, SdkError<E>>`
+- `async fn fetch_existing_client(&self, connector_key: &ConnectorKey) -> Option<Arc<aws_smithy_client::Client>>`
+- `async fn initialize_and_store_new_client<E>(&self, connector_key: &ConnectorKey) -> Result<Arc<aws_smithy_client::Client>, SdkError<E>>`
```
Comment (jdisanti) [inline design/src/rfcs/rfc0015_user_friendly_configurable_connectors.md:102]: It seems like these are essentially private functions needed to implement `get_or_create_client`. Is that correct?

Comment (Velfi) [reply design/src/rfcs/rfc0015_user_friendly_configurable_connectors.md:102]: Yeah, that's correct. You wouldn't call them directly.

**Pull request closed without merging**